nixos/modules/services/web-apps/privatebin.nix

```diff
@@ -0,0 +1,274 @@
+    };
+  };
+
+  configFile = pkgs.writeTextDir "conf.php" ''
+    ;<?php http_response_code(403); /*
+    ${toIni finalCfg}
+    ;*/
+  '';
+
+  mkPhpIni = generators.toKeyValue {
+    mkKeyValue = generators.mkKeyValueDefault { } " = ";
+  };
+  mkPhpPackage = cfg: cfg.phpPackage.buildEnv {
+    extraConfig = mkPhpIni cfg.phpOptions;
+  };
+
+  autoDb = if !cfg.databaseSetup.enable then null else cfg.databaseSetup.kind;
+  db_config = optionalAttrs (autoDb != false) (
+    if autoDb == "mysql" then {
+      model.class = "Database";
+      model_options = {
+        dsn = "mysql:unix_socket=/run/mysqld/mysqld.sock;dbname=${user}";
+        usr = user;
+      };
+    } else if autoDb == "pgsql" then {
+      model.class = "Database";
+      model_options.dsn = "pgsql:host=/run/postgresql;port=${toString config.services.postgresql.port};dbname=${user}";
+    } else if autoDb == "sqlite" then {
+      model.class = "Database";
+      model_options.dsn = "sqlite:/var/lib/privatebin/data/db.sqlite3";
+    } else {}
+  );
+  finalCfg = recursiveUpdate cfg.settings db_config;
+in
+{
+  meta.maintainers = with lib.maintainers; [ e1mo ];
+
+  options.services.privatebin = {
+    enable = mkEnableOption (mdDoc "PrivateBin web application");
+
+    package = mkPackageOptionMD pkgs "privatebin" { };
+
+    settings = mkOption {
+      default = {};
+      example = lib.literalExpression ''
+        {
+          main = {
+            name._env = "PRIVATEBIN_NAME";
+            basepath = "https://privatebin.example.com/";
+            fileupload = true;
+            syntaxhighlightingtheme = "sons-of-obsidian";
+            info = "This instance of PrivateBin is hosted on NixOS!";
+            languageselection = true;
+            icon = "none";
+            cspheader = "default-src 'none'; base-uri 'self'; form-action 'none'; manifest-src 'self'; connect-src * blob:; script-src 'self' 'unsafe-eval'; style-src 'self'; font-src 'self'; frame-ancestors 'none'; img-src 'self' data: blob:; media-src blob:; object-src blob:; sandbox allow-same-origin allow-scripts allow-forms allow-popups allow-modals allow-downloads";
+            httpwarning = true;
+          };
+          expire.default = "1day";
+          traffic = {
+            limit = 10;
+            exempted = "1.2.3.4,10.10.10/24";
+          };
+          purge = {
+            limit = 300;
+            batchsize = 10;
+          };
+          model = {
+            class = "Database";
+          };
+          model_options = {
+            dsn = "pgsql:host=localhost;dbname=privatebin";
+            tbl = "privatebin_";
+            user = "privatebin";
+            pwd._env = "PRIVATEBIN_DB_PASS";
+          };
+        }
+      '';
+      description = lib.mdDoc ''
+        Privatebin configuration as outlined in
+        <https://github.com/PrivateBin/PrivateBin/wiki/Configuration>.
+
+        Available sections are `[main]`, `[expire]`, `[expire_options]`,
```

Comment:
Some of these sections (e.g. expire_options) depend on ordering of the keys. As this is currently just an attribute set (and therefore not ordered) to get a list like this
![image](https://github.com/NixOS/nixpkgs/assets/34243578/6b1ca240-ed7a-441b-8525-1c66628f644a)
one has to make their attribute set look like this
![image](https://github.com/NixOS/nixpkgs/assets/34243578/faaf608a-1526-4d2a-b538-7e1c22995561)

(as PrivateBin labels for expire options parse this name you can't name them something else that sorts correctly alphabetically)

Perhaps we should consider changing this so we may take a list to keep order - although this may be challenging given the nature of the default ini convertor (I'm not sure)

Reply:
Thanks for catching this. Looking at the `toINI` implementation, I am not sure if having ordering by using lists would be possible.
Even with `lib.mkOrder` we'd basically need to re-implement the toINI function[^1].

I think this would at least warrant a discussion in upstream PrivateBin (aka. see if we could leave the ordering to PB). But for the time being we could strip out the `expire_options`, either have it as a list OR manually sort it by expiry time (and move 0 to the end) and then manually perform the INI conversion.

I'd personally prefer the later option (always sorting by the stated time) since the configuration scheme is the same as it's for upstream.

[^1]: The conversion from the attrset into individual lines (aka. the list where we need to look out for ordering) is done in the `toKeyValue` function, which can't be overwritten. See [`generators.nix:129`](https://github.com/NixOS/nixpkgs/blob/b154d4de029bc947300f95f802246b533b6c83a1/lib/generators.nix#L129)